pkg/controller/direct/apikeys/apikeyskey_controller.go

```diff
@@ -0,0 +1,267 @@
+	}
+
+	// if m.config.Endpoint != "" {
+	// 	opts = append(opts, option.WithEndpoint(m.config.Endpoint))
+	// }
+
+	gcpClient, err := api.NewRESTClient(ctx, opts...)
+	if err != nil {
+		return nil, fmt.Errorf("building apikeys client: %w", err)
+	}
+	return gcpClient, err
+}
+
+func (m *model) AdapterForObject(ctx context.Context, u *unstructured.Unstructured) (directbase.Adapter, error) {
+	gcp, err := m.client(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: Just fetch this object?
+	obj := &krm.APIKeysKey{}
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(u.Object, &obj); err != nil {
+		return nil, fmt.Errorf("error converting to %T: %w", obj, err)
+	}
+
+	projectID := obj.Spec.ProjectRef.External
+	if projectID == "" {
+		return nil, fmt.Errorf("unable to determine project")
+	}
+
+	// TODO: Use name or request resourceID to be set on create?
+	keyID := ValueOf(obj.Spec.ResourceID)
+	if keyID == "" {
+		return nil, fmt.Errorf("unable to determine resourceID")
+	}
+
+	location := "global"
+
+	return &adapter{
+		projectID: projectID,
+		location:  location,
+		keyID:     keyID,
+		desired:   obj,
+		gcp:       gcp,
+	}, nil
+}
+
+func (a *adapter) Find(ctx context.Context) (bool, error) {
+	if a.keyID == "" {
+		return false, nil
+	}
+
+	req := &pb.GetKeyRequest{
+		Name: a.fullyQualifiedName(),
+	}
+	key, err := a.gcp.GetKey(ctx, req)
+	if err != nil {
+		if IsNotFound(err) {
+			klog.Warningf("key was not found: %v", err)
+			return false, nil
+		}
+		return false, err
+	}
+
+	u := &krm.APIKeysKey{}
+	if err := keyMapping.Map(key, u); err != nil {
+		return false, err
+	}
+	a.actual = u
+
+	return true, nil
+}
+
+func (a *adapter) Delete(ctx context.Context) (bool, error) {
+	// TODO: Delete via status selfLink?
+	req := &pb.DeleteKeyRequest{
+		Name: a.fullyQualifiedName(),
+	}
+	op, err := a.gcp.DeleteKey(ctx, req)
+	if err != nil {
+		if IsNotFound(err) {
+			return false, nil
+		}
+		return false, fmt.Errorf("deleting key: %w", err)
+	}
+
+	if _, err := op.Wait(ctx); err != nil {
+		return false, fmt.Errorf("waiting for key deletion to complete: %w", err)
+	}
+
+	return true, nil
+}
+
+func (a *adapter) buildCreateRequest(ctx context.Context) (*pb.CreateKeyRequest, error) {
+	// You can configure only the `display_name`, `restrictions`, and
+	// `annotations` fields.
+	desired := &pb.Key{}
+	if err := keyMapping.Map(a.desired, desired); err != nil {
+		return nil, err
+	}
+
+	return &pb.CreateKeyRequest{
+		Parent: fmt.Sprintf("projects/%s/locations/%s", a.projectID, a.location),
+		KeyId:  a.keyID,
+		Key:    desired,
+	}, nil
+}
+
+func (a *adapter) Create(ctx context.Context, u *unstructured.Unstructured) error {
+	req, err := a.buildCreateRequest(ctx)
+	if err != nil {
+		return err
+	}
+
+	op, err := a.gcp.CreateKey(ctx, req)
+	if err != nil {
+		return fmt.Errorf("creating key: %w", err)
+	}
+	// TODO: Is the resourceID returned if it is dynamically created?  Maybe we should create the UUID?
```

Comment:
I think AIPs say it should be returned in the initial LRO and the "completed" LRO; though that doesn't mean it is.

IIRC, if we don't specify a name here, the API will server-generate one.  We probably want to avoid that!

However, if we didn't just use the metadata.name, we could instead pre-generate a name, put it in the spec.resourceID and go from there.  But better to use the metadata.name I think.

This code is good enough for our existing tests.  But I need more tests so that we can make these decisions.  There are (at least) two ways to add more tests, one is to add more testfixtures for APIKeysKey, the other is to add more direct-coded controllers so we get more testfixtures that way.  I plan on doing both.